read_slice() / write_slice() methods to permit dataset slice IO

pmarks opened the pull request:
Add a reader method that supports fetching ndarray slices out of a dataset, letting the user avoid reading the entire dataset. Any feedback on the general approach?  If it seems like we're on the right track here I'll 1-d and 2-d special cases and writer support.

TODOs:
 - [x] support SliceOrIndex::Index() case
 - [x] special case 1D and 2D sliced reads
 - [x] writer support

Enet4 commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,6 +54,69 @@ impl Deref for Dataspace {
+    T: AsRef<[SliceOrIndex]>, D: ndarray::Dimension
+    {
+        let shape = self.dims();
+        let mut start_vec = Vec::new();
```
These vectors can be allocated with the expected capacity.

```suggestion
        let mut start_vec = Vec::with_capacity(ss.len());
```

Enet4 commented on hdf5-rs/tests/common/gen.rs:
```diff
@@ -17,6 +17,55 @@ pub fn gen_ascii<R: Rng + ?Sized>(rng: &mut R, len: usize) -> String {
+/// Generate a random slice of elements inside the given `shape` dimension.
+pub fn gen_slice<R: Rng + ?Sized>(rng: &mut R, shape: &[usize]) -> Vec<SliceOrIndex> {
+
+    let mut r = Vec::new();
```
`Iterator` all the things:

```rust
shape.into_iter()
    .map(|s| gen_slice_one_dim(rng, *s))
    .collect()
```

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -35,19 +36,56 @@ impl<'a> Reader<'a> {
+    /// dataset.  FIXME - ensure this: Sliced reads are not allowed from attributes.
+    /// Use the multidimsional slice macro `s![]` from `ndarray` to conveniently create
+    /// a multidimensional slice.
+    pub fn read_slice<T,S,D>(&self, slice: SliceInfo<S,D>) -> Result<Array<T, D>>
```
Shouldn't the slice be a ref?

Looking at [`ArrayBase::slice()`](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#method.slice), it takes a `&SliceInfo<...>`?

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -35,19 +36,56 @@ impl<'a> Reader<'a> {
+            let fspace = self.obj.space()?;
+            let out_shape = fspace.select_slice(slice)?;
+            let mspace = Dataspace::create_simple(&out_shape)?;
+            let size = out_shape.iter().fold(1 ,|p, v| p * (*v));
```
There's `iter().product()` (sometimes the type inference fails and you will have to specify its return type via turbofish).

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -35,19 +36,56 @@ impl<'a> Reader<'a> {
+        if shape.ndim() == 0 {
+            // Fall back to a simple read for the scalar case
+            // Slicing has no effect
+            self.read()
```
It's an edge case, but - should this fail loudly? E.g. it would fail in h5py, I think.

pmarks replied:
I think this is consistent with h5py. There's is only one 0-dimensional slice (the empty array), and when applied to a scalar dataset it just returns the value:

```
$ipython
In [13]: f = h5py.File("test.h5", "w")

In [14]: f.create_dataset("t1", (), data=5)
Out[14]: <HDF5 dataset "t1": shape (), type "<i8">

In [15]: tt = f["t1"][()]

In [16]: tt
Out[16]: 5
```

aldanor replied:
Yea but your slice dimensions could be anything here? (because you're checking `obj.shape()`)

If you try this in h5py, it should fail:

```python
>>> f['t1'][1:2]
```

pmarks replied:
you're correct -- this is now by the `if shape.ndim() != slice_dim {` check above.

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -35,19 +36,56 @@ impl<'a> Reader<'a> {
+
+            self.read_into_buf(vec.as_mut_ptr(), Some((&fspace, &mspace))).unwrap();
+
+            let arr = ArrayD::from_shape_vec(out_shape, vec).str_err().unwrap();
```
No unwraps please?

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -35,19 +36,56 @@ impl<'a> Reader<'a> {
+                vec.set_len(size);
+            }
+
+            self.read_into_buf(vec.as_mut_ptr(), Some((&fspace, &mspace))).unwrap();
```
No unwraps?

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -1,9 +1,14 @@
 use std::fmt::{self, Debug};
 use std::ops::Deref;
 use std::ptr;
+use ndarray::{SliceInfo, SliceOrIndex};
```
The imports in this crate are usually split like this:

```rust
use std::foo;
use std::bar;

use third_party1::foo;
use third_party2::bar;

use hdf5_rs_workspace_stuff;

use crate::local::stuff;
```

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,6 +54,69 @@ impl Deref for Dataspace {
 }
 
 impl Dataspace {
+
```
Extra newline

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,6 +54,69 @@ impl Deref for Dataspace {
 impl Dataspace {
+
+    /// Create a simple dataspace covering the given `shape`
+    pub fn create_simple(shape: &[Ix]) -> Result<Dataspace> {
```
How is this different from `Dataspace::try_new()?` (aside from a better name perhaps) Or should it at least reuse that?

(dataspaces/shapes/dims will likely be heavily refactored in the future anyway to integrate better with ndarray)

pmarks replied:
sorry, missed try_new. Got rid of this method.

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,6 +54,69 @@ impl Deref for Dataspace {
 impl Dataspace {
+
+    /// Create a simple dataspace covering the given `shape`
+    pub fn create_simple(shape: &[Ix]) -> Result<Dataspace> {
```
`Result<Dataspace>` -> `Result<Self>`, it's slightly nicer / more flexible.

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,6 +54,69 @@ impl Deref for Dataspace {
+    /// Select a slice (known as a 'hyperslab' in HDF5 terminology) of the Dataspace.
+    /// Returns the shape of array that is capable of holding the resulting slice.
+    /// Useful when you want to read a subset of a dataset.
+    pub fn select_slice<T, D>(&self, slice: SliceInfo<T,D>) -> Result<Vec<Ix>> where
```
`slice` can be taken by ref?

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,6 +54,69 @@ impl Deref for Dataspace {
+            SliceOrIndex::Slice { start, end, step } => {
+                let end = end.unwrap_or(*shape as isize);
+                if end > (*shape as isize) {
+                    panic!("invalid selection");
```
No panics please - this should return a `Result<>`

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,6 +54,69 @@ impl Deref for Dataspace {
+                }
+
+                if *step < 1 {
+                    panic!("Step must be >= 1 (got {})", step);
```
No panics

aldanor commented:
To leave some comments on the general approach I'd need to read some official HDF5 docs on hyperslabs, will get back later with that :)

pmarks commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -35,19 +36,56 @@ impl<'a> Reader<'a> {
     }
 
-    fn read_into_buf<T: H5Type>(&self, buf: *mut T) -> Result<()> {
+    fn read_into_buf<T: H5Type>(&self, buf: *mut T, file_space: Option<(&Dataspace, &Dataspace)>) -> Result<()> {
```
agreed

pmarks commented:
@aldanor @Enet4  Thanks for the quick reviews! I've made all the suggested changes expect for the point about slicing a scalar dataset.

Re the selelct_hyperslab pattern: from past work I have the impression that this is a very common style.  This setup is very similar to how h5py does things. Here's a C example: https://support.hdfgroup.org/ftp/HDF5/current/src/unpacked/examples/h5_subset.c

I'll keep going with this general style unless there are additional style ideas.

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,7 +53,66 @@ impl Deref for Dataspace {
+            SliceOrIndex::Slice { start, end, step } => {
+                let end = end.unwrap_or(*shape as isize);
+                if end > (*shape as isize) {
+                    return Err(Error::Internal("slice is larger than dataset".to_string()));
```
You could just do `fail!("slice is larger than the dataset")` (or "slice extends beyond dataspace bounds" since this is about dataspace).

aldanor replied:
In fact you can even do an `ensure!(end <= shape as _,"slice extends beyond dataspace bounds")`.

pmarks replied:
yes, done.

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,7 +53,66 @@ impl Deref for Dataspace {
+        Ok(shape_vec)
+    }
+
+    fn get_start_stride_count(v: &SliceOrIndex, shape: &Ix) -> Result<(u64, u64, u64)> {
```
No point in taking `&Ix` by reference here since it's a scalar, can just do an `Ix`.

Also, `shape` is a misleading name here? Maybe `len` would have ben better?

pmarks replied:
good call, fixed.

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,7 +53,66 @@ impl Deref for Dataspace {
+                    return Err(Error::Internal("slice is larger than dataset".to_string()));
+                }
+
+                if *step < 1 {
```
`ensure!(*step >= 1, "step must be >= 1 (got {})", step)`

pmarks replied:
yes, thanks

aldanor commented on hdf5-rs/src/hl/space.rs:
```diff
@@ -49,7 +53,66 @@ impl Deref for Dataspace {
+                    return Err(Error::Internal(format!("Step must be >= 1 (got {})", step)));
+                }
+
+                if end < *start {
```
This whole thing (including logic below) probably simplifies to something like this?

```rust
1 + (end - start - 1).max(0) / step
```

The corner case is when `end == start` -- wondering why do you return `count=1` when `end < start`, but `count=0` when `end == start`?
```

pmarks replied:
> ```rust
> 1 + (end - start - 1).max(0) / step
> ```
This will give `count = 1` when `end == start`, which I don't want.

> The corner case is when `end == start` -- wondering why do you return `count=1` when `end < start`, but `count=0` when `end == start`?

`end < start` is handled above. I changed the comparison to emphasize that `count=0` for empty intervals.  `1 + (end - start - 1).max(0) / step`

pmarks commented:
@aldanor OK, I've added 1-d and 2-d readers, a general writer, added some additional checks, and generalized it to support 'single index' `SliceOrIndex::Index` slices, which cause the dimensionality of the output of the slice to change. Tests are passing on stable but not nightly, which looks like a compiler bug to me.

Anything else you want me to tackle here?

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -91,6 +92,19 @@ impl<'a> Reader<'a> {
+            let reduced_shape: Vec<usize> = slice_s
+                .iter()
+                .zip(out_shape.iter().cloned())
+                .filter(|(slc, sz)| match slc {
```
Minor: I think this filter + map could be combined:

```rust
.filter_map(|(slc, sz)| match slc {
    SliceOrIndex::Index(_) => None,
    _ => Some(sz)
})
```

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -261,18 +277,41 @@ impl<'a> Writer<'a> {
+            let mspace = Dataspace::try_new(&slice_shape, false)?;
 
             // FIXME - we can handle non-standard input arrays by creating a memory space
             // that reflects the same slicing/ordering that this ArrayView represents.
```
Wondering - how hard would it be to support non standard layouts? As long as it's convertible to a hyperslab spec, it should be fine, no?

I'm thinking maybe all this "hyperslab" <--> "ndarray indexing" business should be moved out into a separate API to simplify all this logic here (and tested separately, without even touching HDF5 stuff). Not saying it should be done in this PR, just generally/later. Thoughts?

pmarks replied:
yes, definitely - should be doable, and should be abstracted out of this code. Filed as #21.

aldanor commented on hdf5-rs/tests/test_dataset.rs:
```diff
@@ -67,6 +67,9 @@ where
 
         // Ensure that the H5 sliced read matches the ndarray slice of the original array.
+        if sliced_read != sliced_dataset {
+            println!("{:?}", slice);
```
`assert_eq!()`?

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -223,11 +239,9 @@ impl<'a> Writer<'a> {
     }
 
-
     /// Reads a slice of an n-dimensional array.
```
"Writes"? Should probably be also clarified what exactly is being sliced here.

aldanor commented on hdf5-rs/src/hl/container.rs:
```diff
@@ -261,18 +277,41 @@ impl<'a> Writer<'a> {
+            for s in slice_s {
+                match s {
+                    SliceOrIndex::Index(_) => {
+                        data_shape_hydrated.push(1);
```
1 or 0?

pmarks replied:
1 -- SliceOrIndex::Index take a single element from that dimension.

aldanor commented:
@pmarks I've updated the feature/2018 branch with some of the recent changes (including fixes to the CI scripts so hopefully random failures will go away) -- could you please rebase your branch?

Maybe also squash most/all of your commits as well as you do it?

pmarks commented:
Ok, everything is rebased & squashed. fmt clean & no warnings.  Thanks!

aldanor commented:
@pmarks Thanks!

(There's stuff to clean up and refactor, but we'll figure it later)
